src/helpers/authentication.js

```diff
@@ -85,30 +85,37 @@ module.exports = (application) => {
   };
 
   const refreshToken = (user) => {
-    if (user === 'default') return;
     if (isTest) return;
 
     clearTimeout(timeouts[user]);
 
     const token = tokens[user];
     const createdAt = toMilliseconds(token.createdAt);
     const expiresIn = toMilliseconds(tokens[user].expiresIn);
-    const expiresAt = createdAt + expiresIn;
+    const expiresAt = createdAt + expiresIn - 600_000; // refresh 10 minutes before expected expiry
 
-    timeouts[user] = setTimeout(async () => {
-      config.logger.debug(`[API][${application} Starting Refreshing Token`);
+    config.logger.debug(
+      `[API][${application}] Token will be refreshed at ${new Date(expiresAt)}.`
+    );
 
-      const response = await fetchToken({
-        grantType: 'refresh_token',
-        refreshToken: tokens[user].refreshToken,
-      });
+    timeouts[user] = setTimeout(async () => {
+      config.logger.debug(`[API][${application}] Starting Refreshing Token`);
+
+      const response = await fetchToken(
```

Comment:
When the request fails the `fetchToken` will raise an error but I don't see how this would be handled? Would it bouble up and kill the process?

Reply:
yes, it will bubble up and kill the process.. all unhandled promises kill the process..

Reply:
Is that the desired behavior though?

Reply:
why not? without valid token, the middleware can not function.. so, it's okay to let it die (and be restarted by docker swarm).. nginx would still able respond to the webhook with an error status..

not sure if we can gracefully handle it any other way.. :thinking:

Reply:
But the token is a problem of a single request, not the entire service. So eventually that user's request should fail and return an appropriate error message, right? Even if the error message is 'please contact an admin to resolve this'.

Reply:
Depends, default laboperator token is used by all users..

The webhook coming from server come with an encoded token, decoding that token requires a valid default token.. so, if the default token failed to be refreshed, it affect all users, and at that point it makes no differences if the process is killed, or if it's handled with custom error message..

The IAM tokens are per user, so in that case, you're right.. failure to refresh the token of a single users potentially should not affect other users.. but, then again, I'm not sure in what situation would refreshing tokens failed just for a single users, but not the others? Considering that server would first need to have successfully authenticate the users with the IAM provider, before it can pass along the authentication to the middleware..

Separately, I've also been wondering if refreshing the tokens asynchronously based on `setTimeout` is a good strategy or not.. alternatively, we could do it lazily in the middleware request handler, e.g.

1. Request came in with encoded token
2. If default token is expired, refresh default token
3. Decode token
4. If user IAM token is expired, refresh IAM token
5. Use token to fetch data from the 3rd party

wdyt?

Reply:
I had also thought about that. Would also be more appropriate in terms of any activity logging that happens on the IAM side.

Reply:
OK, since it will require bigger changes, I file a new [JIRA bug ticket](https://labforward.atlassian.net/browse/CUB-2595)..

I'll close this PR and will open a new one..